filebeat/input/kafka/input.go

```diff
@@ -0,0 +1,211 @@
+// under the License.
+
+package kafka
+
+import (
+	"context"
+	"time"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/elastic/beats/filebeat/channel"
+	"github.com/elastic/beats/filebeat/input"
+	"github.com/elastic/beats/filebeat/util"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/pkg/errors"
+)
+
+func init() {
+	err := input.Register("kafka", NewInput)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Input contains the input and its config
+type Input struct {
+	config        kafkaInputConfig
+	rawConfig     *common.Config // The Config given to NewInput
+	started       bool
+	outlet        channel.Outleter
+	consumerGroup sarama.ConsumerGroup
+	kafkaContext  context.Context
+	kafkaCancel   context.CancelFunc // The CancelFunc for kafkaContext
+	log           *logp.Logger
+}
+
+// NewInput creates a new kafka input
+func NewInput(
+	cfg *common.Config,
+	outletFactory channel.Connector,
+	inputContext input.Context,
+) (input.Input, error) {
+
+	out, err := outletFactory(cfg, inputContext.DynamicFields)
+	if err != nil {
+		return nil, err
+	}
+
+	config := defaultConfig
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "reading kafka input config")
+	}
+
+	saramaConfig, err := newSaramaConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing Sarama config")
+	}
+	consumerGroup, err :=
+		sarama.NewConsumerGroup(config.Hosts, config.GroupID, saramaConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing kafka consumer group")
+	}
+
+	// Sarama uses standard go contexts to control cancellation, so we need to
+	// wrap our input context channel in that interface.
+	kafkaContext, kafkaCancel := context.WithCancel(context.Background())
+	go func() {
+		select {
+		case <-inputContext.Done:
+			logp.Info("Closing kafka context because input stopped.")
+			kafkaCancel()
+			return
+		}
+	}()
+
+	input := &Input{
+		config:        config,
+		rawConfig:     cfg,
+		started:       false,
+		outlet:        out,
+		consumerGroup: consumerGroup,
+		kafkaContext:  kafkaContext,
+		kafkaCancel:   kafkaCancel,
+		log:           logp.NewLogger("kafka input").With("hosts", config.Hosts),
+	}
+
+	return input, nil
+}
+
+func (p *Input) newConsumerGroup() (sarama.ConsumerGroup, error) {
+	consumerGroup, err :=
+		sarama.NewConsumerGroup(p.config.Hosts, p.config.GroupID, nil)
+	return consumerGroup, err
+}
+
+// Run starts the input by scanning for incoming messages and errors.
+func (p *Input) Run() {
+	if !p.started {
+		// Track errors
+		go func() {
+			for err := range p.consumerGroup.Errors() {
+				p.log.Errorw("Error reading from kafka", "error", err)
+			}
+		}()
+
+		go func() {
+			for {
+				handler := groupHandler{input: p}
+
+				err := p.consumerGroup.Consume(p.kafkaContext, p.config.Topics, handler)
+				if err != nil {
+					p.log.Errorw("Kafka consume error", "error", err)
+				}
```

Comment:
I don't believe so. Errors calling `Consume` should mean failure to join the consumer group to begin with, so there's no state to recover and all we can do is retry.

Reply:
Why can clients fail to join a consumer group? Permissions, network failures, topic does not exist (yet)?

Reply:
The errors that can be returned here are:

```
ErrClosedConsumerGroup
ErrInvalidTopic
ErrTopicAuthorizationFailed
ErrUnknownTopicOrPartition
ErrLeaderNotAvailable
ErrOutOfBrokers
ErrConsumerCoordinatorNotAvailable
```

(as far as I can tell there's no centralized list of possible errors in either the docs or the code, so I might have missed one, but this is everything I know of)

Reply:
This is interesting, and how this actually interacts with the filebeat runner.
Currently the runner executes `Run` once. If Run returns and `--once` CLI flag is set by the user, then filebeat will return. But `--once` is meant to collect and stop once it reached EOF. Here `--once` forces a shutdown if you encountered an error in `Consume`.
If the user did not configure `--once`, then filebeat will wait for `scan_frequency` seconds (yep, your kafka input has a `scan_frequency` setting), and call `Run` again. Some inputs start a another go-routine. This is why they guard themselves from extra calls to Run. This doesn't play well with these inputs, cause filebeat and the runner think that the input has finished if `--once` is set, but this is not true...

That is, only by chance filebeat will retry if `Consume` fails here. Yet I'd like to remove this behavior from the runner (which behavior is specific/correct for the logs input only), or have an alternative runner.

If an input is statically configured, but returns/fails, then filebeat (normally) should shut down. This is also the case if your constructor fails. e.g. NewConsumerGroup might fail, and therefore force a filebeat shutdown if no other input is configured. With filebeat being a service, it should retry (with backoff) if an error is recoverable/healable.
For the errors given I'd say all can be resolved while filebeat is running, as all seem to be triggered by the outside. e.g. `ErrLeaderNotAvailable`, `ErrConsumerCoordinatorNotAvailable`, `ErrConsumerCoordinatorNotAvailable`, or `ErrConsumerCoordinatorNotAvailable` might indicate network errors or kafka cluster being unstable. `ErrInvalidTopic` might eventually resolved once a producer comes online. `ErrUnknownTopicOrPartition` is not helpful. While a topic might be unknown, it might get generated. Having the wrong partition might be more severe, but this should only occur if someone removed partitions.

Question: when can NewConsumerGroup error? I guess if the config is wrong, but also if connection attempt fails? In the former case we should return, but the later one would be 'recoverable'.

Reply:
NewConsumerGroup only errors on configuration as far as I know, since it doesn't actually establish the connection it only creates the interface for it. The consumer group isn't actually joined until calling Consume.

In any case, alright, I will guard against multiple calls for now :-)

Reply:
I have to amend this after more testing, the initial call to `NewConsumerGroup` in `NewInput` does actively join the group and will return failure if it can't reach the brokers. In light of that, I think it is better to wait to call `NewConsumerGroup` until `input.Run`, so our retry loop includes the initial connection, otherwise an ephemeral network error on startup could require a restart of the beat... oops.